### ISSUE TITLE
Add support for ECS Task Roles

### ISSUE THREAD

**Issue by jonjitsu** [deliciousbrains/wp-amazon-s3-and-cloudfront#415]
Currently the plugin doesn't support ECS Task Roles. This seems to be a limitation based on it's dependency on the wp-amazon-web-services plugin that uses sdk v2.

**Comment by ianmjones** [comment]
Thanks for the feature request @jonjitsu.

**Comment by wdegeus** [comment]
Any update on this? Currently running into this on dockerized wordpress installations, we don't want to give EC2 blanket permissions to S3 since those permissions will be too broad, and creating an IAM user per site sucks, I don't want useless users of which we have to toss around the secrets :)

**Comment by ianmjones** [comment]
We need to switch to AWS PHP SDK v3 first, make sure that's working well, and then we can look at implementing features like this that require SDK v3.

**Comment by wdegeus** [comment]
Any idea on how long that switch will take? Any beta version we can try?

I would hope that switching to the v3 SDK automagically makes this work, since it's basically the same as using an IAM role for EC2, except for it's a different role. AWS already sets some magic environment variables in the docker container for handling the ECS task role. (and the aws cli works just fine with that).
See https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html

**Comment by ianmjones** [comment]
@wdegeus We don't give out projected release dates publicly as they often change due to unforeseen issues or shifting priorities which only leads to disappointment.

**Comment by Veraxus** [comment]
I’m running into this too. It’s a pretty major security flaw, IMO. We figured we could maybe use Composer to pull a newer version, but while the AWS PHP SDK is bundled with the plugin in the vendors directory, it’s not an actual dependency - and you appear to have modified the library’s name spacing for some reason. ☹️

**Comment by ianmjones** [comment]
@Veraxus please see https://deliciousbrains.com/wp-offload-s3-1-6-released/ for why and how we bundled what we need of the AWS PHP SDK.

**Comment by rv-aburdine** [comment]
I believe this is fixed since this plugin now uses aws sdk v3. I've used it in a ecs task and it works for me.

**Comment by ianmjones** [comment]
That's great to hear @rv-aburdine.

Did you have to do anything special or specifically related to WP Offload S3 to "hook it up"?

**Comment by rv-aburdine** [comment]
here's a snippet of the wp-config I used that worked:

```php
define('AS3CF_AWS_USE_EC2_IAM_ROLE', true);
define('WPOS3_SETTINGS', serialize([
  'bucket' => 'bucket-name',
  'region' => 'us-east-1',
  'copy-to-s3' => true,
  'serve-from-s3' => false,
  'enable-object-prefix' => true,
  'object-prefix' => 'wp-content/uploads/',
  'use-yearmonth-folders' => true,
  'remove-local-file' => true,
  'domain' => 'cloudfront',
  'cloudfront' => $siteUrl,
  'object-versioning' => false
]));
```

that, plus the plugin files copied into a wordpress docker container running on ECS with the task role set, worked like a charm.

**Comment by ianmjones** [comment]
Thanks @rv-aburdine. It looks like the existing method of ...

`define('AS3CF_AWS_USE_EC2_IAM_ROLE', true);`

... is all that's needed now that WP Offload S3 is using AWS PHP SDK v3.

I'm closing this issue as fixed, but if anyone using WP Offload S3 1.7.2+ or WP Offload S3 Lite 1.4.3+ finds that ECS Task Roles aren't being properly handled, please comment as we'll open up again for investigation.
